### ISSUE TITLE
卷积神经网络（Convolutional Neural Network, CNN）算法概览

### ISSUE THREAD

**Issue by nyarray** [nyarray/archives#7]
> 概览卷积神经网络的主要概念、常用结构、训练与优化技巧、常见变体与工程实践，并包含最小可运行的 PyTorch / TensorFlow 示例代码和调优建议，适合作为 README 或教学笔记直接放入项目中。

---

## 1. 简介

卷积神经网络（CNN）是处理网格结构数据（如图像）的深度学习模型。核心思想：用可学习的卷积核（filters）在空间上提取局部特征，并通过逐层组合形成高层语义表示。CNN 在图像分类、目标检测、语义分割、姿态估计等任务上取得了巨大成功。

### 卷积层（Convolutional Layer）

* 作用：用若干卷积核对输入进行局部加权求和，提取特征图（feature maps）。
* 重要参数：核大小（kernel size）、步幅（stride）、填充（padding）、输出通道数（out channels）。
* 公式（单通道、无偏置示意）：

$$
Y(i,j)=\sum_{u=0}^{k-1}\sum_{v=0}^{k-1}X(i+u, j+v)\cdot W(u,v)
$$

* 常见技巧：`same` padding 保持尺寸；深度可分离卷积（Depthwise Separable Conv）用于加速与轻量化。

### 激活函数

* ReLU（常用）: $f(x)=\max(0,x)$
* Leaky ReLU / ELU / GELU：应对 ReLU 死亡神经元问题。
* Sigmoid/Tanh：已较少用于中间层，可能用于二分类输出。

### 池化层（Pooling）

* 最大池化（MaxPool）和平均池化（AvgPool）。
* 作用：下采样、降低空间尺寸、提高感受野、引入平移不变性。

### 批归一化（Batch Normalization）

* 缓解内部协变量偏移，加速收敛，常放在卷积/线性层与激活之间或之后。

### Dropout

* 随机失活神经元，减少过拟合。图像任务中卷积层上常用率较低（如 0.1–0.3），全连接层上常用 0.5。

### 全连接层（Fully Connected, FC）

* 用于高层语义整合与最终分类。现代架构常减少 FC 层，更多依赖 GAP（Global Average Pooling）。

## 3. 常见网络结构简述

* **LeNet**：早期简单结构，适合手写数字识别。
* **AlexNet**：用 ReLU、Dropout、数据增强推动深度学习复兴。
* **VGG**：大量 3x3 卷积堆叠，结构简洁但参数多。
* **ResNet**：引入残差块 `y = F(x) + x`，解决深层网络退化问题。
* **Inception**：多分支（1x1、3x3、5x5）并行提取不同感受野的特征。
* **MobileNet / EfficientNet**：针对移动/嵌入式场景的轻量化网络（深度可分离卷积、复合缩放策略）。

## 4. 损失函数与优化器

* 分类：交叉熵（Cross-Entropy）($\mathcal{L}_{CE}$)。
* 回归：均方误差（MSE）。
* 优化器：SGD（带动量）、Adam、AdamW（Adam + 权重衰减修正）。
* 学习率调度：StepLR、Cosine Annealing、ReduceLROnPlateau、Warmup 等。

## 5. 正则化与数据增强

* 数据增强：随机裁剪、翻转、旋转、色彩抖动、MixUp、CutMix。
* 权重衰减（Weight Decay）等价于 L2 正则化。
* 早停（Early Stopping）、标签平滑（Label Smoothing）。

## 6. 训练技巧与工程实践

* 初始化：He initialization（ReLU 网络）或 Xavier（Glorot）。
* 批大小（batch size）与学习率成正比（线性缩放规则）。
* 使用混合精度训练（FP16）加速与节省显存。
* 分布式训练：数据并行（DataParallel / DDP）。
* 模型检查点（checkpoint）和训练日志（TensorBoard / WandB）。
* 使用 `warmup` 阶段防止初期学习率过大导致发散。

## 7. 可视化与解释性

* 特征图（feature map）可视化：观察卷积层中激活。
* Grad-CAM、Grad-CAM++：基于梯度的类激活映射，解释模型关注区域。

## 8. 实践示例

下面给出最小可运行的 PyTorch 和 TensorFlow（Keras）示例。

### PyTorch 最小 CNN（训练骨架）

```python
# minimal_cnn_pytorch.py
import torch
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader

class SimpleCNN(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1),
            nn.BatchNorm2d(32),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2),

            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2),
        )
        self.classifier = nn.Sequential(
            nn.AdaptiveAvgPool2d((1,1)),
            nn.Flatten(),
            nn.Linear(64, num_classes)
        )

    def forward(self, x):
        x = self.features(x)
        x = self.classifier(x)
        return x

# 数据与训练循环省略：使用 CIFAR10 数据集与常见 training loop

```

### TensorFlow / Keras 最小示例

```python
# minimal_cnn_tf.py
import tensorflow as tf
from tensorflow.keras import layers, models

def build_simple_cnn(input_shape=(32,32,3), num_classes=10):
    model = models.Sequential([
        layers.Conv2D(32, 3, padding='same', activation='relu', input_shape=input_shape),
        layers.BatchNormalization(),
        layers.MaxPool2D(),

        layers.Conv2D(64, 3, padding='same', activation='relu'),
        layers.BatchNormalization(),
        layers.MaxPool2D(),

        layers.GlobalAveragePooling2D(),
        layers.Dense(num_classes, activation='softmax')
    ])
    return model

# 编译与训练省略：model.compile(...); model.fit(...)
```

## 9. 常见问题与故障排查

* **训练不收敛**：检查学习率、初始化、数据预处理（是否归一化）、损失函数是否合理。
* **过拟合**：增强数据、加大正则化、降低模型容量或使用早停。
* **显存不足**：减小 batch size、使用混合精度、使用梯度累积。
* **训练速度慢**：确保使用 GPU、优化数据加载（num_workers）、使用 AMP。

## 10. 参考与延伸阅读

* 各经典网络论文（LeNet、AlexNet、VGG、ResNet、Inception、MobileNet）
* 深度学习框架官方文档（PyTorch、TensorFlow / Keras）
* 进阶主题：目标检测（Faster R-CNN、YOLO 系列）、语义分割（U-Net、DeepLab）
